core/input/input.cpp

```diff
@@ -944,6 +1323,160 @@ void Input::vibrate_handheld(int p_duration_ms, float p_amplitude) {
 	OS::get_singleton()->vibrate_handheld(p_duration_ms, p_amplitude);
 }
 
+#define CHECK_TRIGGER                                                                                      \
+	if (!joy_names.has(p_device) && p_axis != JoyAxis::TRIGGER_LEFT && p_axis != JoyAxis::TRIGGER_RIGHT) { \
+		return;                                                                                            \
+	}
+#define SETUP_TRIGGER_EFFECT(m_mode)                                                                                                        \
+	int trigger_id = p_axis == JoyAxis::TRIGGER_LEFT ? 0 : 1;                                                                               \
+                                                                                                                                            \
+	ScePadTriggerEffectParam effect = {};                                                                                                   \
+	ScePadTriggerEffectCommand &command = effect.command[trigger_id];                                                                       \
+                                                                                                                                            \
+	effect.triggerMask = p_axis == JoyAxis::TRIGGER_LEFT ? SCE_PAD_TRIGGER_EFFECT_TRIGGER_MASK_L2 : SCE_PAD_TRIGGER_EFFECT_TRIGGER_MASK_R2; \
+	command.mode = m_mode;
+#define SEND_EFFECT \
+	JoypadSDL::get_singleton()->send_effect(p_device, &effect, sizeof(effect));
+
+void Input::joy_adaptive_triggers_off(int p_device, JoyAxis p_axis) {
+	_THREAD_SAFE_METHOD_
+	CHECK_TRIGGER;
+	SETUP_TRIGGER_EFFECT(SCE_PAD_TRIGGER_EFFECT_MODE_OFF);
+	SEND_EFFECT;
+}
+
+void Input::joy_adaptive_triggers_feedback(int p_device, JoyAxis p_axis, int p_position, int p_strength) {
+	_THREAD_SAFE_METHOD_
+	CHECK_TRIGGER;
+
+	if (p_position < 0 || p_position > 9 || p_strength < 0 || p_strength > 8) {
+		return;
+	}
+
+	SETUP_TRIGGER_EFFECT(SCE_PAD_TRIGGER_EFFECT_MODE_FEEDBACK);
+	command.commandData.feedbackParam.position = p_position;
+	command.commandData.feedbackParam.strength = p_strength;
+
+	SEND_EFFECT;
+}
+
+void Input::joy_adaptive_triggers_weapon(int p_device, JoyAxis p_axis, int p_start_position, int p_end_position, int p_strength) {
+	_THREAD_SAFE_METHOD_
+	CHECK_TRIGGER;
+
+	if (p_start_position < 2 || p_start_position > 7) {
+		return;
+	}
+	if (p_end_position <= p_start_position || p_end_position > 8) {
+		return;
+	}
+	if (p_strength < 0 || p_strength > 8) {
+		return;
+	}
```

Comment:
Invalid parameters should print an error with `ERR_FAIL_COND_MSG()` instead of failing silently. (Same for all other adaptive trigger methods.)